Core: HMR for StoryIndex server

tmeasday opened the pull request:
Issue: #16047

## What I did

- [x] Watch each stories entry for changes with `watchpack` and invalidate the `StoryIndexGenerator`'s internal cache when changes happen
- [x] Added a SSE endpoint for `stories.json` that emits an `INVALIDATE` event when the list changes
- [x] Create a `StoryIndexClient` class that opens a `EventSource` on the SSE endpoint
- [x] Reload the manager + preview story index when the client emits.

### What I still need to do
 - [x] Some questions about naming + where to put common files
 - [x] Properly deal with non-recursive specifiers => https://github.com/storybookjs/storybook/pull/16220
 - [x] Consider not generating dynamic imports in `to-importFn` (later)
 - [x] Need to double check behaviour works in all scenarioes
 - [ ] Some questions about efficiency
 - [x] Add tests etc

tmeasday committed:
First pass working

shilman commented on lib/api/src/modules/stories.ts:
```diff
@@ -496,6 +500,8 @@ export const init: ModuleFn = ({
     );
 
+    indexClient = new StoryIndexClient();
+    indexClient.addEventListener('INVALIDATE', () => fullAPI.fetchStoryList());
```
can you explain the flow of control here?

tmeasday replied:
The server code detects a change, triggers the invalidation event. The client code picks up the event, triggering the fetch stories index code to refetch it.

shilman commented on lib/core-server/src/utils/stories-json.ts:
```diff
@@ -1,6 +1,10 @@
 import { StoryIndexGenerator } from './StoryIndexGenerator';
+import { watchStorySpecifiers } from './watch-story-specifier';
+
+const eventName = 'INVALIDATE';
```
```suggestion
const INVALIDATE = 'INVALIDATE';
```

shilman commented on lib/core-server/src/utils/stories-json.ts:
```diff
@@ -15,21 +19,51 @@ export async function extractStoriesJson(
+      let closed = false;
+      const watcher = () => {
+        if (closed || res.writableEnded) return;
+        res.write(`event:INVALIDATE\ndata:DATA\n\n`);
```
```suggestion
        res.write(`event:${INVALIDATE}\ndata:DATA\n\n`);
```

shilman replied:
what's `DATA`?

tmeasday replied:
Oh yeah whoops. Thanks. Took my a while to get this to work

shilman commented on lib/core-server/src/utils/watch-story-specifier.ts:
```diff
@@ -0,0 +1,47 @@
+        !!path.startsWith(base.replace(/^\.\//, '')),
+        !!path.match(new RegExp(regex))
+      );
+      return path.startsWith(base.replace(/^\.\//, '')) && path.match(new RegExp(regex));
```
This `./` seems a little dicey? Would be nice to figure this out globally somehow so we don't have to keep special casing things everywhere

tmeasday replied:
I agree. Consolidating the specifier code (with the require context and import fn code also) is at the top of my list here.

shilman commented on lib/preview-web/src/StoryIndexClient.ts:
```diff
@@ -0,0 +1,18 @@
+
+const PATH = './stories.json';
+
+export class StoryIndexClient extends EventSource {
```
duplicate to above? not sure i understand why we need this anyway. is it for testing purposes?

tmeasday replied:
I need to figure out where a common spot is this should live. We need it for both the manager and the preview. 

Not sure what you mean by testing

shilman reviewed (commented):
@tmeasday would be useful to walk through this together. i understand the basic premise but don't understand where exactly the events ping-pong around to and when

tmeasday committed:
Tightend up backend code and added tests

tmeasday committed:
Fix up some more types to bootstrap

tmeasday committed:
Add stories for manager HMR

tmeasday committed:
Get HMR working for v6&v7

tmeasday committed:
Little tweak to get generation working again properly

We had an issue with absolute vs relative paths. Going to sort this out properly when I handle all specifiers soon.

tmeasday committed:
Fix up store tests

tmeasday committed:
Fixed up `PreviewWeb` tests

tmeasday committed:
Un-skip two tests we couldn't implement before

tmeasday committed:
Add a bunch of edge case tests to `PreviewWeb`

tmeasday committed:
First pass at unified specifiers working

tmeasday committed:
Fix `autoTitle` test

tmeasday committed:
Fixed up `normalize-stories` test

tmeasday committed:
Fix up require-context tests

tmeasday committed:
Added tests for `watch-story-specifiers`

tmeasday committed:
Fix `StoryIndexGenerator` tests

tmeasday committed:
Add integration test for `stories-json`

tmeasday committed:
Some fixes

tmeasday committed:
Linting cleanup

tmeasday committed:
Require context is generated relative to `workingDir` now

tmeasday committed:
Fix some tests

tmeasday committed:
Use an 0.7 canary of IE addon to get `EventSource` support

tmeasday committed:
Update ie11 addon again to unshift

tmeasday committed:
Trying again with the plugin

tmeasday committed:
Update snapshots

tmeasday committed:
We don't use event source in this test but we inherit from it

tmeasday committed:
Merge remote-tracking branch 'origin/next' into 16047-refactor-specifiers

tmeasday committed:
Update snapshots w/ back compat story index

tmeasday committed:
Directory test happens before we make glob relative to workingDir

tmeasday committed:
Don't start watching stories.json dependencies until someone requests it.

This mainly to stop tests from hanging

shilman commented on lib/core-server/src/utils/StoryIndexGenerator.ts:
```diff
@@ -44,10 +48,19 @@ export class StoryIndexGenerator {
+        const pathToSubIndex = {} as SpecifierStoriesCache;
 
         const files = await glob(path.join(this.configDir, specifier.glob));
+        console.log('specifiers');
```
console.log

shilman reviewed (commented):
This is looking great. Biggest concern is all the console.logs left over :D

Should I review the telescoping PR too?

tmeasday commented:
Yes! I think the console logs are removed there.

tmeasday committed:
Merge remote-tracking branch 'origin/next' into 16047-add-hmr-to-story-index-2

tmeasday committed:
Merge branch '16047-add-hmr-to-story-index-2' into 16047-refactor-specifiers

tmeasday committed:
Update addons/storyshots/storyshots-core/src/api/index.ts

tmeasday committed:
Pull `storiesV2Compatibility` into `options`

shilman committed:
Merge pull request #16220 from storybookjs/16047-refactor-specifiers

Core: Unified story specifiers

shilman committed:
update from next

shilman commented:
@tmeasday try the following. first run `react-ts` in one window:

```
cd examples/react-ts
yarn storybook
```

Open up your browser dev tools.

Then in another window:

```
cd examples/react-ts/src
mv emoji-button.stories.js foo-button.stories.js
```

This doesn't always fail, but if you change the filename, you should see errors like this in the browser console, and the UI will freeze/die.

<img width="664" alt="Examples___Emoji_Button_-_With_Args_⋅_Storybook" src="https://user-images.githubusercontent.com/488689/136324842-0238106c-edc3-4b03-be54-c823121b6df8.png">

shilman committed:
Merge branch 'next' into 16047-add-hmr-to-story-index-2

shilman committed:
Fix failing tests

shilman committed:
Fix deepscan
